IDR/notebook/scripts/nbconverted/2.collect_statistics.py

```diff
@@ -0,0 +1,354 @@
+    )
+    h, pi_list = h_index(p=rel_frequencies)
+
+    # Calculate Normalized Median Evenness
+    nme = norm_median_evenness(pi_list)
+
+    # Calculate Pielou's evenness
+    j = pielou(h=h, s=s)
+
+    # Calculate Gini coefficient
+    gc = gini_coef(abs_frequencies)
+
+    return s, h, nme, j, gc
+
+
+# In[18]:
+
+
+def collect_study_stats(
+    metadata_file_path,
+    results_list,
+    na_cols=[
+        "screen_id",
+        "study_name",
+        "plate_name",
+        "plate_id",
+        "sample",
+        "pixel_size_x",
+        "pixel_size_y",
+    ],
+):
+    """Collecting statistics within a single file
+
+    Parameters
+    ----------
+    metadata_file_path: PosixPath object
+        Path to single study .parquet metadata file
+    results_list: list
+        Outside instantiated list to append study statistics
+    na_cols: list
+        Image attributes excluded from statistical calculations
+
+    Returns
+    -------
+    results_list: list
+        The input results list appended with statistics from the metadata_file_path study
+    """
+    # Read parquet into pandas df
+    metadata_df = pd.read_parquet(metadata_file_path)
+
+    # Extract metadata from file name and dataframe
+    metadata_pq = metadata_file_path.name
+    study_name = metadata_pq.split(".")[0]
+    attribute_names = metadata_df.columns.to_list()
+
+    # Remove irrelevant attributes
+    for attribute in na_cols:
+        attribute_names.remove(attribute)
+
+    # Collect statistics for each attribute
+    for attribute in attribute_names:
+        unique_entries = metadata_df[attribute].unique()
+        attribute_elements = dict()
+        for element in unique_entries:
+            attribute_elements[element] = len(
+                metadata_df[metadata_df[attribute] == element]
+            )
+
+        s, h, nme, j, gc = stats_pipeline(attribute_elements=attribute_elements)
+
+        # Append stats to attribute_results
+        results_list.append([study_name, attribute, s, h, nme, j, gc])
+
+    return results_list
+
+
+# In[19]:
+
+
+# Define study metadata directory
+studies_metadata_dir = pathlib.Path("../data/metadata")
+
+# Collect metadata file paths
+metadata_files = list(walk(studies_metadata_dir))
+
+# Calculate statistics for each image attribute within each study
+all_results_list = list()
+for metadata_path in metadata_files:
+    collect_study_stats(metadata_path, all_results_list)
+
+stat_results_df = pd.DataFrame(
+    data=all_results_list,
+    columns=["Study_Name", "Attribute", "S", "H", "NME", "J", "GC"],
+)
+
+# Make directories
+stats_dir = pathlib.Path("../data/statistics")
+pathlib.Path.mkdir(stats_dir, exist_ok=True)
+
+# Save individual stats as parquet file
+output_file = pathlib.Path(stats_dir, f"individual_studies_diversity.parquet.gzip")
+stat_results_df.to_parquet(output_file, compression="gzip")
+
+
+# In[20]:
+
+
+def collect_databank_stats(metadata_dir, na_cols=["pixel_size_x", "pixel_size_y"]):
```

Comment:
functions should be defined prior to running code. Consider moving this function up!

Also, b/c you are going to move this to production eventually (correct? I can see this notebook actually doing all the heavy lifting here so maybe not) maybe you should consider moving these functions into a `scripts/stat_util.py` file and importing here.

Reply:
The `collect_databank_stats()` function was moved before the code begins to run.

Added in 02e7d3225894c69fd949343f8fda2155de0b11ab.

Reply:
@gwaybio @axiomcura I've already moved many of these "utils-like" functions to a `utils` sub-directory in production for a separate branch. I'll do the same for the notebook folder.

Reply:
Utils-like functions were moved to `scripts/stat_utils.py` and were imported into `2.collect_statistics,py`.

Added in 7c67bf55c7e93a92b7c7c56eebb82f34d33bf781.